### PR TITLE
Support expressions with window functions

### LINKED ISSUE #19857: SUM(col) / SUM(SUM(col)) OVER (PARTITION BY col2) aggregate function over WINDOW is not supported.
**How to reproduce**

```
SELECT sum(number) / sum(sum(number)) OVER (PARTITION BY (number % 10))
FROM numbers(10000)
GROUP BY number % 10

Query id: 76ed1093-b378-45c2-b644-1de44783734a


0 rows in set. Elapsed: 0.003 sec.

Received exception from server (version 21.2.1):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Unknown identifier: sum(sum(number)) OVER (PARTITION BY number % 10) there are columns: modulo(number, 10), sum(number): While processing sum(number) / sum(sum(number)) OVER (PARTITION BY (number % 10)).



postgres=# SELECT sum(generate_series) / sum(sum(generate_series))  OVER (PARTITION BY (generate_series % 10)) FROM generate_series(0,10000) GROUP BY generate_series % 10;
        ?column?
------------------------
 1.00000000000000000000
 1.00000000000000000000
 1.00000000000000000000
 1.00000000000000000000
 1.00000000000000000000
 1.00000000000000000000
 1.00000000000000000000
 1.00000000000000000000
 1.00000000000000000000
 1.00000000000000000000
(10 rows)

**Expected behavior**
Query works

### PR DESCRIPTION
**Pull request by novikd**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Support expressions with window functions. Closes #19857

### CODE REVIEW

**Comment by novikd** [comment]
@Mergifyio update

File: src/Interpreters/InDepthNodeVisitor.h [hunk 1]
```diff
@@ -95,4 +95,33 @@ class OneTypeMatcher
 template <typename Data, NeedChild::Condition need_child = NeedChild::all>
 using ConstOneTypeMatcher = OneTypeMatcher<Data, need_child, const ASTPtr>;
 
+template <typename Visitor, typename T = ASTPtr>
+struct InDepthNodeVisitorWithChildInfo : Visitor
+{
+    using ChildInfo = typename Visitor::ChildInfo;
```
Comment (vdimir) [inline src/Interpreters/InDepthNodeVisitor.h:101]: Why do we need special visitor for this, can't we use regular one and store the `info` in the `data`?

File: src/Interpreters/ActionsVisitor.h [hunk 2]
```diff
@@ -120,6 +121,12 @@ class ActionsMatcher
 public:
     using Visitor = ConstInDepthNodeVisitor<ActionsMatcher, true>;
 
+    enum class WindowDependancyState
```
Comment (vdimir) [inline src/Interpreters/ActionsVisitor.h:124]: Why not just `bool depends_on_window` ?

File: src/Interpreters/ActionsVisitor.cpp [hunk 3]
```diff
@@ -945,12 +963,45 @@ void ActionsMatcher::visit(const ASTFunction & node, const ASTPtr & ast, Data &
         // Don't need to do anything more for window functions here -- the
         // resulting column is added in ExpressionAnalyzer, similar to the
         // aggregate functions.
+        if (data.window_dependancy_state == WindowDependancyState::MAY_DEPEND)
+            data.window_function_in_subtree = true;
         return;
     }
+    else if (node.compute_after_window_functions)
```
Comment (vdimir) [inline src/Interpreters/ActionsVisitor.cpp:970]: So, this flags `subtree_contains_window_call`, `compute_after_window_functions`, `window_function_in_subtree`, `window_dependancy_state` are required just to change the order of visiting the nodes, are they? It looks like this PR's trickiest part, but it just does `visit`.

File: src/Interpreters/ExpressionAnalyzer.cpp [hunk 4]
```diff
@@ -1388,6 +1409,15 @@ void SelectQueryExpressionAnalyzer::appendWindowFunctionsArguments(
     }
 }
 
+void SelectQueryExpressionAnalyzer::appendExpressionsAfterWindowFunctions(ExpressionActionsChain & chain, bool /* only_types */)
+{
+    ExpressionActionsChain::Step & step = chain.lastStep(columns_after_window);
+    for (const auto & expression : syntax->expressions_with_window_function)
+    {
+        getRootActionsForWindowFunctions(expression->clone(), true, step.actions());
```
Comment (vdimir) [inline src/Interpreters/ExpressionAnalyzer.cpp:1417]: So, generally, most of the work is done here. Have I got right that it's the part that drives the calculation of expressions for window functions?

File: src/Parsers/ASTFunction.h [hunk 5]
```diff
@@ -22,6 +22,8 @@ class ASTFunction : public ASTWithAlias
 
     bool is_window_function = false;
 
+    bool compute_after_window_functions = false;
```
Comment (vdimir) [inline src/Parsers/ASTFunction.h:25]: AST Looks like not the best place for this flag, but we already have `is_window_function` here, so it seems now there's no easy way to get rid of it. And we need actually to determine which expressions are related to the window in `WindowExpressionsCollectorChildInfo` that also operates with AST, though

File: src/Interpreters/ActionsVisitor.cpp (hunk 3 above)
Comment (novikd) [reply src/Interpreters/ActionsVisitor.cpp:970]: Yes, that's the most important part of this PR. Let's consider some general expression with a window function:
```
f(g(arg) OVER window_name)
```
Evaluation of this expression must be split to 3 separate parts:
1. Evaluate all arguments of window functions (`arg` in the example)
2. Evaluate window functions. It's important to notice, that any window function can't have another window function as an argument. (In the example it's `g(arg) OVER window_name`)
3. Evaluate expressions with window functions (in the example we will compute`f` call value). We can do this way because we are sure that `f` is not:
    * Aggregate function (aggregation is executed before window functions, so I can't have a window function as an argument);
    * Window function.

In other words, here we have some kind of _topological sort_.

File: src/Interpreters/ActionsVisitor.h (hunk 2 above)
Comment (novikd) [reply src/Interpreters/ActionsVisitor.h:124]: It's not affecting performance in any way, but IMHO makes code more readable. This variable describes the current state of the visitor, so enum represents it more straightforwardly than boolean.

File: src/Interpreters/ExpressionAnalyzer.cpp (hunk 4 above)
Comment (novikd) [reply src/Interpreters/ExpressionAnalyzer.cpp:1417]: Yes, here we create actions for expression evaluation.

File: src/Interpreters/InDepthNodeVisitor.h (hunk 1 above)
Comment (novikd) [reply src/Interpreters/InDepthNodeVisitor.h:101]: I want explicitly show that:
1. Node processing result depends on the results of children nodes.
2. Children are processed completely independently. `Data` represents global visitor values.

File: src/Parsers/ASTFunction.h (hunk 5 above)
Comment (novikd) [reply src/Parsers/ASTFunction.h:25]: In my opinion, it's completely okay to cache some useful data in AST. Instead of that we usually recompute the same values several times.

The only part that should be considered is that we don't want to make AST nodes too big. In this case, it's not a problem, because I just added a simple boolean flag (in the opposite situation we could store a pointer to a struct with cached values).

File: src/Interpreters/InDepthNodeVisitor.h (hunk 1 above)
Comment (novikd) [reply src/Interpreters/InDepthNodeVisitor.h:101]: For example, `GetAggregatesVisitor` could be implemented in mode efficient way using this approach.
Instead of traversing query AST for several times in `getAggregates` and `getWindowFunctions` functions, we can collect functions and check correctness on the fly if we know independent results for child nodes.

**Review by vdimir — APPROVED** [review]

**Comment by novikd** [comment]
@Mergifyio update

**Comment by novikd** [comment]
> Stateless tests (release, s3 storage, actions) — fail: 2, passed: 4148, skipped: 53

Unrelated.

**Comment by vpanfilov** [comment]
@novikd, thanks for working on this functionality.
I tested this feature on unstable version 22.7.1.710. Simple expressions like:
```sql
SELECT sum(number) OVER() + 1 AS result FROM numbers(10);

SELECT arraySum(groupArray(number) OVER()) AS result FROM numbers(10);

SELECT sum(number) OVER() + avg(number) OVER() AS result FROM numbers(10);
```
now works.

But it is not working for array operations with lambda expressions:
```sql
SELECT arrayMap(x -> x + 1, groupArray(number) OVER()) AS result FROM numbers(10);

Code: 183. DB::Exception: Unexpected lambda expression: While processing arrayMap(x -> (x + 1), groupArray(number) OVER ()) AS result. (UNEXPECTED_EXPRESSION) (version 22.7.1.710 (official build))
```

And not working in case of nested window functions:
```sql
SELECT sum(avg(number) OVER()) OVER() AS result FROM numbers(10);

Code: 184. DB::Exception: Window function avg(number) OVER () is found inside another window function in query: While processing avg(number) OVER (). (ILLEGAL_AGGREGATION) (version 22.7.1.710 (official build))
```

**Comment by UnamedRus** [comment]
> And not working in case of nested window functions:

It's possible to have nested WINDOW functions?

**Comment by novikd** [comment]
@vpanfilov Thank you for testing the feature. I'll take a look at higher-order functions.

> And not working in case of nested window functions:

SQL standard defines nested window functions as a completely different entity. The query you wrote is not supposed to work. If you want, you can use GROUP BY to fix this query.

Nested window functions specified in the SQL standard require specific <row_marker> usage. As far as I know, it's not been supported in ClickHouse yet.
